File: services/entity-resolution-service/src/fusion_base/domain/merge.rs
use std::collections::BTreeMap;

use chrono::Utc;
use serde_json::{Map, Value};
use uuid::Uuid;

use crate::models::{
    cluster::{EntityRecord, ResolvedCluster},
    golden_record::{GoldenRecord, GoldenRecordProvenance},
    merge_strategy::{MergeStrategy, SurvivorshipRule},
};

use super::engine::comparator::normalize_text;

pub fn synthesize_golden_record(
    cluster: &ResolvedCluster,
    strategy: &MergeStrategy,
) -> GoldenRecord {
    let now = Utc::now();
    let field_rules = if strategy.rules.is_empty() {
        default_survivorship_rules()
    } else {
        strategy.rules.clone()
    };

    let mut canonical_values = Map::new();
    let mut provenance = Vec::new();

    for rule in &field_rules {
        if let Some(selected) = select_value(&cluster.records, rule, &strategy.default_strategy) {
            canonical_values.insert(rule.field.clone(), selected.value.clone());
            provenance.push(GoldenRecordProvenance {
                field: rule.field.clone(),
                source: selected.source,
                external_id: selected.external_id,
                strategy: selected.strategy,
            });
        }
    }

    if !canonical_values.contains_key("display_name") {
        if let Some(record) = cluster.records.first() {
            canonical_values.insert(
                "display_name".to_string(),
                Value::String(record.display_name.clone()),
            );
        }
    }

    let title = canonical_values
        .get("display_name")
        .or_else(|| canonical_values.get("name"))
        .and_then(|value| value.as_str())
        .unwrap_or("Golden Record")
        .to_string();

    let completeness_score =
        (canonical_values.len() as f32 / field_rules.len().max(1) as f32).clamp(0.0, 1.0);

    GoldenRecord {
        id: Uuid::now_v7(),
        cluster_id: cluster.id,
        title,
        canonical_values: Value::Object(canonical_values),
        provenance,
        completeness_score,
        confidence_score: cluster.confidence_score,
        status: if cluster.status == "rejected" {
            "rejected"
        } else {
            "active"
        }
        .to_string(),
        created_at: now,
        updated_at: now,
    }
}

#[derive(Debug)]
struct SelectedValue {
    value: Value,
    source: String,
    external_id: String,
    strategy: String,
}

fn select_value(
    records: &[EntityRecord],
    rule: &SurvivorshipRule,
    default_strategy: &str,
) -> Option<SelectedValue> {
    let strategy = if rule.strategy.trim().is_empty() {
        default_strategy.to_string()
    } else {
        rule.strategy.clone()
    };

    match strategy.as_str() {
        "source_priority" => select_source_priority(records, rule),
        "highest_confidence" => select_highest_confidence(records, &rule.field, &strategy),
        "most_common" => select_most_common(records, &rule.field, &strategy),
        _ => select_longest_non_empty(records, &rule.field, &strategy),
    }
}

fn select_source_priority(
    records: &[EntityRecord],
    rule: &SurvivorshipRule,
) -> Option<SelectedValue> {
    for source in &rule.source_priority {
        for record in records {
            if &record.source != source {
                continue;
            }

            if let Some(value) = extract_value(record, &rule.field) {
                return Some(SelectedValue {
                    value,
                    source: record.source.clone(),
                    external_id: record.external_id.clone(),
                    strategy: "source_priority".to_string(),
                });
            }
        }
    }

    select_longest_non_empty(records, &rule.field, &rule.fallback)
}

fn select_highest_confidence(
    records: &[EntityRecord],
    field: &str,
    strategy: &str,
) -> Option<SelectedValue> {
    records
        .iter()
        .filter_map(|record| extract_value(record, field).map(|value| (record, value)))
        .max_by(|left, right| left.0.confidence.total_cmp(&right.0.confidence))
        .map(|(record, value)| SelectedValue {
            value,
            source: record.source.clone(),
            external_id: record.external_id.clone(),
            strategy: strategy.to_string(),
        })
}

fn select_longest_non_empty(
    records: &[EntityRecord],
    field: &str,
    strategy: &str,
) -> Option<SelectedValue> {
    records
        .iter()
        .filter_map(|record| extract_value(record, field).map(|value| (record, value)))
        .max_by(|left, right| value_length(&left.1).cmp(&value_length(&right.1)))
        .map(|(record, value)| SelectedValue {
            value,
            source: record.source.clone(),
            external_id: record.external_id.clone(),
            strategy: strategy.to_string(),
        })
}

fn select_most_common(
    records: &[EntityRecord],
    field: &str,
    strategy: &str,
) -> Option<SelectedValue> {
    let mut counts = BTreeMap::<String, (usize, &EntityRecord, Value)>::new();

    for record in records {
        let Some(value) = extract_value(record, field) else {
            continue;
        };
        let normalized = normalize_text(value.as_str().unwrap_or_default());
        let entry = counts
            .entry(normalized)
            .or_insert((0, record, value.clone()));
        entry.0 += 1;
    }

    counts
        .into_values()
        .max_by(|left, right| left.0.cmp(&right.0))
        .map(|(_, record, value)| SelectedValue {
            value,
            source: record.source.clone(),
            external_id: record.external_id.clone(),
            strategy: strategy.to_string(),
        })
}

fn extract_value(record: &EntityRecord, field: &str) -> Option<Value> {
    match field {
        "display_name" | "name" => Some(Value::String(record.display_name.clone())),
        _ => record.attributes.get(field).cloned().and_then(|value| {
            if value.is_null() {
                None
            } else if value.as_str().is_some_and(|value| value.trim().is_empty()) {
                None
            } else {
                Some(value)
            }
        }),
    }
}

fn value_length(value: &Value) -> usize {
    value.as_str().map(str::len).unwrap_or(0)
}

fn default_survivorship_rules() -> Vec<SurvivorshipRule> {
    vec![
        SurvivorshipRule {
            field: "display_name".to_string(),
            strategy: "longest_non_empty".to_string(),
            source_priority: vec!["crm".to_string(), "erp".to_string(), "support".to_string()],
            fallback: "highest_confidence".to_string(),
        },
        SurvivorshipRule {
            field: "email".to_string(),
            strategy: "source_priority".to_string(),
            source_priority: vec!["crm".to_string(), "erp".to_string(), "support".to_string()],
            fallback: "most_common".to_string(),
        },
        SurvivorshipRule {
            field: "phone".to_string(),
            strategy: "most_common".to_string(),
            source_priority: vec![],
            fallback: "longest_non_empty".to_string(),
        },
        SurvivorshipRule {
            field: "company".to_string(),
            strategy: "most_common".to_string(),
            source_priority: vec![],
            fallback: "longest_non_empty".to_string(),
        },
    ]
}

File: openfoundry-go/services/entity-resolution-service/internal/domain/merge.go
package domain

import (
	"sort"
	"strings"
	"time"

	"github.com/openfoundry/openfoundry-go/services/entity-resolution-service/internal/domain/engine"
	"github.com/openfoundry/openfoundry-go/services/entity-resolution-service/internal/models"
)

// SynthesizeGoldenRecord ports `merge::synthesize_golden_record`.
func SynthesizeGoldenRecord(cluster models.ResolvedCluster, strategy models.MergeStrategy) models.GoldenRecord {
	now := time.Now().UTC()

	fieldRules := strategy.Rules
	if len(fieldRules) == 0 {
		fieldRules = defaultSurvivorshipRules()
	}

	canonicalValues := map[string]any{}
	provenance := []models.GoldenRecordProvenance{}

	for _, rule := range fieldRules {
		selected, ok := selectValue(cluster.Records, rule, strategy.DefaultStrategy)
		if !ok {
			continue
		}
		canonicalValues[rule.Field] = selected.value
		provenance = append(provenance, models.GoldenRecordProvenance{
			Field:      rule.Field,
			Source:     selected.source,
			ExternalID: selected.externalID,
			Strategy:   selected.strategy,
		})
	}

	if _, ok := canonicalValues["display_name"]; !ok {
		if len(cluster.Records) > 0 {
			canonicalValues["display_name"] = cluster.Records[0].DisplayName
		}
	}

	title := "Golden Record"
	if v, ok := canonicalValues["display_name"]; ok {
		if s, ok := v.(string); ok {
			title = s
		}
	} else if v, ok := canonicalValues["name"]; ok {
		if s, ok := v.(string); ok {
			title = s
		}
	}

	denominator := len(fieldRules)
	if denominator < 1 {
		denominator = 1
	}
	completeness := float32(len(canonicalValues)) / float32(denominator)
	switch {
	case completeness < 0:
		completeness = 0
	case completeness > 1:
		completeness = 1
	}

	status := "active"
	if cluster.Status == "rejected" {
		status = "rejected"
	}

	return models.GoldenRecord{
		ID:                engine.MustNewUUIDv7(),
		ClusterID:         cluster.ID,
		Title:             title,
		CanonicalValues:   canonicalValues,
		Provenance:        provenance,
		CompletenessScore: completeness,
		ConfidenceScore:   cluster.ConfidenceScore,
		Status:            status,
		CreatedAt:         now,
		UpdatedAt:         now,
	}
}

type selectedValue struct {
	value      any
	source     string
	externalID string
	strategy   string
}

func selectValue(records []models.EntityRecord, rule models.SurvivorshipRule, defaultStrategy string) (selectedValue, bool) {
	strategy := strings.TrimSpace(rule.Strategy)
	if strategy == "" {
		strategy = defaultStrategy
	}
	switch strategy {
	case "source_priority":
		return selectSourcePriority(records, rule)
	case "highest_confidence":
		return selectHighestConfidence(records, rule.Field, strategy)
	case "most_common":
		return selectMostCommon(records, rule.Field, strategy)
	default:
		return selectLongestNonEmpty(records, rule.Field, strategy)
	}
}

func selectSourcePriority(records []models.EntityRecord, rule models.SurvivorshipRule) (selectedValue, bool) {
	for _, source := range rule.SourcePriority {
		for _, record := range records {
			if record.Source != source {
				continue
			}
			value, ok := extractValue(record, rule.Field)
			if !ok {
				continue
			}
			return selectedValue{
				value:      value,
				source:     record.Source,
				externalID: record.ExternalID,
				strategy:   "source_priority",
			}, true
		}
	}
	return selectLongestNonEmpty(records, rule.Field, rule.Fallback)
}

func selectHighestConfidence(records []models.EntityRecord, field, strategy string) (selectedValue, bool) {
	var best *models.EntityRecord
	var bestValue any
	for i := range records {
		value, ok := extractValue(records[i], field)
		if !ok {
			continue
		}
		if best == nil || records[i].Confidence > best.Confidence {
			best = &records[i]
			bestValue = value
		}
	}
	if best == nil {
		return selectedValue{}, false
	}
	return selectedValue{
		value:      bestValue,
		source:     best.Source,
		externalID: best.ExternalID,
		strategy:   strategy,
	}, true
}

func selectLongestNonEmpty(records []models.EntityRecord, field, strategy string) (selectedValue, bool) {
	var best *models.EntityRecord
	var bestValue any
	bestLen := -1
	for i := range records {
		value, ok := extractValue(records[i], field)
		if !ok {
			continue
		}
		l := valueLength(value)
		if l > bestLen {
			best = &records[i]
			bestValue = value
			bestLen = l
		}
	}
	if best == nil {
		return selectedValue{}, false
	}
	return selectedValue{
		value:      bestValue,
		source:     best.Source,
		externalID: best.ExternalID,
		strategy:   strategy,
	}, true
}

func selectMostCommon(records []models.EntityRecord, field, strategy string) (selectedValue, bool) {
	type bucket struct {
		count    int
		record   *models.EntityRecord
		value    any
		insertAt int
	}
	counts := map[string]*bucket{}
	insertion := 0
	for i := range records {
		value, ok := extractValue(records[i], field)
		if !ok {
			continue
		}
		var s string
		if str, isString := value.(string); isString {
			s = str
		}
		key := engine.NormalizeText(s)
		if existing, ok := counts[key]; ok {
			existing.count++
			continue
		}
		counts[key] = &bucket{count: 1, record: &records[i], value: value, insertAt: insertion}
		insertion++
	}
	if len(counts) == 0 {
		return selectedValue{}, false
	}
	// Stable max — Rust uses BTreeMap iter order (key-sorted) + max_by
	// (returns the last max). We replicate by iterating sorted keys.
	keys := make([]string, 0, len(counts))
	for k := range counts {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	var winner *bucket
	for _, k := range keys {
		b := counts[k]
		if winner == nil || b.count >= winner.count {
			winner = b
		}
	}
	return selectedValue{
		value:      winner.value,
		source:     winner.record.Source,
		externalID: winner.record.ExternalID,
		strategy:   strategy,
	}, true
}

func extractValue(record models.EntityRecord, field string) (any, bool) {
	switch field {
	case "display_name", "name":
		return record.DisplayName, true
	default:
		v, ok := record.Attributes[field]
		if !ok {
			return nil, false
		}
		if v == nil {
			return nil, false
		}
		if s, isStr := v.(string); isStr && strings.TrimSpace(s) == "" {
			return nil, false
		}
		return v, true
	}
}

func valueLength(value any) int {
	if s, ok := value.(string); ok {
		return len(s)
	}
	return 0
}

func defaultSurvivorshipRules() []models.SurvivorshipRule {
	return []models.SurvivorshipRule{
		{
			Field:          "display_name",
			Strategy:       "longest_non_empty",
			SourcePriority: []string{"crm", "erp", "support"},
			Fallback:       "highest_confidence",
		},
		{
			Field:          "email",
			Strategy:       "source_priority",
			SourcePriority: []string{"crm", "erp", "support"},
			Fallback:       "most_common",
		},
		{
			Field:          "phone",
			Strategy:       "most_common",
			SourcePriority: []string{},
			Fallback:       "longest_non_empty",
		},
		{
			Field:          "company",
			Strategy:       "most_common",
			SourcePriority: []string{},
			Fallback:       "longest_non_empty",
		},
	}
}

